Lesson 20. Мини-фреймворк Simpler

rusdec opened the pull request:
Создайте на гитхабе форк учебного проекта из скринкаста (ссылки в дополнительных материалах), в полученном новом репозитории, в новой ветке выполните следующие задания:
- Реализуйте расширенные возможности метода render которые позволят возвращать ответ в других форматах, например:   `render plain: "Plain text response"`
- Реализуйте возможность устанавливать в методе контроллера статус ответа, например:  `status 201`
- Реализуйте возможность устанавливать в методе контроллера заголовки, например:  `headers['Content-Type'] = 'text/plain'`
- Реализуйте механизм обработки исключения когда маршрут для запрашиваемого URL не был найден. В этом случае клиенту должен отдаваться ответ со статусом 404
- Напишите механизм разбора route-параметров. Например, при добавлении маршрута  
`get '/tests/:id', 'tests#show'`
  - Маршрут должен корректно обрабатывать GET запрос  `/tests/101`
  - В методе show контроллера при вызове метода params должен быть доступен параметр :id со значением 101
- Напишите middleware для логирования HTTP-запросов и ответов:
  - Лог должен записываться в файл log/app.log
  - Для запросов необходимо записывать HTTP-метод запроса, URL, контроллер и метод который будет обрабатывать запрос, хэш параметров который будет доступен при вызове метода params
  - Для ответов необходимо записывать код статуса ответа, тип тела ответа и название шаблона (если ответ рендерился с помощью шаблона представления)
            
Пример лога:
```
Request: GET /tests?category=Backend
Handler: TestsController#index
Parameters: {'category' => 'Backend'}
Response: 200 OK [text/html] tests/index.html.erb
```

Commit:
Задание 1. Расшиеренные возможности метода Controller::render

Реализуйте расширенные возможности метода render которые позволят возвращать ответ в других форматах, например:

- render plain: "Plain text response"

Commit:
Задание 2 и 3
- Реализуйте возможность устанавливать в методе контроллера статус ответа, например: status 201
- Реализуйте возможность устанавливать в методе контроллера заголовки, например: headers['Content-Type'] = 'text/plain'

Commit:
Задание 1 и 3. рефакторинг
- Метод set_content_type устанавливает заголовок Content-Type и одновременно @content_type. Чтобы была синхронизация

Commit:
Задание 4. Статусом 404 если маршрут не найден
- Реализуйте механизм обработки исключения когда маршрут для запрашиваемого URL не был найден. В этом случае клиенту должен отдаваться ответ со статусом 404

Commit:
Задание 4. Рефакторинг

Commit:
Задание 4. Рефакторинг

Commit:
Задание 5. Маршруты с параметрами
- Напишите механизм разбора route-параметров. Например, при добавлении маршрута get '/tests/:id', 'tests#show'
  - Маршрут должен корректно обрабатывать GET запрос /tests/101
  - В методе show контроллера при вызове метода params должен быть доступен параметр :id со значением 101

Commit:
Задание 6. Логирование
Напишите middleware для логирования HTTP-запросов и ответов:
- Лог должен записываться в файл log/app.log
- Для запросов необходимо записывать HTTP-метод запроса, URL, контроллер и метод который будет обрабатывать запрос, хэш параметров который будет доступен при вызове метода params
- Для ответов необходимо записывать код статуса ответа, тип тела ответа и название шаблона (если ответ рендерился с помощью шаблона представления)

psylone commented on lib/simpler/application.rb:
```diff
@@ -28,14 +30,28 @@ def routes(&block)
-      controller = route.controller.new(env)
-      action = route.action
+      if route
+        parse_params(route: route, env: env)
```
Можно полностью сделать логику парсинга параметров в объекте роута.

rusdec replied:
хорошо, посмотрю, как можно изменить

rusdec replied:
Она там и была, но я порефакторил: https://github.com/rusdec/simpler/pull/1/commits/9d1bc30e3fa5896bbfd8b9efcc6041bf9947a5ce

psylone commented on lib/simpler/controller.rb:
```diff
@@ -1,54 +1,125 @@
+    end
+
+    def query_string_params
+      @request.params
```
Это могут быть ещё и параметры POST-запроса. Будет лучше мерджить `@request.params` с роут-параметрами внутри метода `params`.

rusdec replied:
не учитывал. хорошо, изменю

rusdec replied:
Готово. https://github.com/rusdec/simpler/pull/1/commits/9d1bc30e3fa5896bbfd8b9efcc6041bf9947a5ce

psylone commented on lib/simpler/controller.rb:
```diff
@@ -1,54 +1,125 @@
+    end
+
+    def parse_render_data(data)
+      _data = { body: nil, type: :html }
```
Вообще, часто используют `_` если переменная не используется. В Go и Elixir особенно. Можно так оставить, но я бы как-то иначе назвал один из параметров.

rusdec replied:
знаю. в го так и делал. Тут я их использую для показа "временности" переменной, то есть она тут просто чтобы похранить значения и не более того. Принял во внимание, посмотрю стуле-руби-гайд на этот момент

rusdec replied:
[Нашёл в РРС](https://github.com/bbatsov/ruby-style-guide#underscore-unused-vars)
> Prefix with _ unused block parameters and local variables. It's also acceptable to use just _ (although it's a bit less descriptive). This convention is recognized by the Ruby interpreter and tools like RuboCop and will suppress their unused variable warnings.

rusdec replied:
Всё же исправил название переменной https://github.com/rusdec/simpler/pull/1/commits/9b4c4a9828c49fc37c97d19822255b5c284b6c3a

psylone commented on lib/simpler/controller.rb:
```diff
@@ -1,54 +1,125 @@
 
+    def set_content_type(type)
+      self.content_type = type
+      set_content_type_header(type)
```
Дублируется инфа, заголовка должно быть достаточно.

rusdec replied:
ммм.... я обязательно посмотрю, как это можно исправить

rusdec replied:
Всё же оказалось недостаточно. Теперь архитектура такова, что View для рендеринга должен понимать, какой тип контента ему рендерить. Для этого нужно устанавливать тип не только в заголовок ответа, но и в env. Так как env доступен для View. Исправлено в https://github.com/rusdec/simpler/pull/1/commits/0129a5b1b1f5cd20cef1f5e81a46e618327c977c

psylone commented on lib/simpler/controller.rb:
```diff
@@ -1,54 +1,125 @@
+      if data[:type] == :html
+        @request.env['simpler.template'] = data[:body]
+      else
+        self.content_body = data[:body]
```
Здесь нарушается интерфейс. В первом случае мы модифицируем `env`, а здесь - записываем сразу тело ответа.

rusdec replied:
да, ведь json, plain-текст отдаются сразу в контроллере:
```
def show
render  json: { @test }
end
```

в то время как html ещё лезет во view и оттуда забирает даные

rusdec replied:
ваш комментарий ниже снял вопрос...

rusdec replied:
Исправлено: https://github.com/rusdec/simpler/pull/1/commits/0129a5b1b1f5cd20cef1f5e81a46e618327c977c

psylone commented on lib/simpler/controller.rb:
```diff
@@ -1,54 +1,125 @@
+    def render(data)
+      data = parse_render_data(data)
+
+      if data[:type] == :html
```
Лучше работать через `case`, ведь могут быть и другие типы ответа: `json`, `xml`. Они у тебя даже есть в допустимых content types.

rusdec replied:
думал об этом так же. Хорошо, изменю

rusdec replied:
Исправлено: https://github.com/rusdec/simpler/pull/1/commits/0129a5b1b1f5cd20cef1f5e81a46e618327c977c

psylone commented on lib/simpler/controller.rb:
```diff
@@ -1,54 +1,125 @@
+      send "render_body_#{content_type}"
+    end
+
+    def render_body_html
```
Вот этого метода здесь по идее быть не должно. контроллер ничего не должен знать о том **как** рендерить ответ. Он только контролирует и управляет ходом запроса.

То есть, должена быть абстракция `View` которой мы говорим: "Хэй, абстракция, отрендери-ка нам вот это в таком-то формате!" И дальше `View` сама должна внутри себя разрулить логику что и как рендерить. Можно сделать отедльные классы для рендеринга HTML, plain text, JSON, XML и прочего.

Также `View` должна уметь работать независимо от контроллера. Создали объект, передали `env`, получили response body.

rusdec replied:
хорошо, изменю

rusdec replied:
Исправлено. Для каждого формата используется свой класс. В классе View происходит инстанциирование нужного класса и вызов у него метода render. Исправлено: https://github.com/rusdec/simpler/pull/1/commits/0129a5b1b1f5cd20cef1f5e81a46e618327c977c

psylone commented on lib/simpler/router/route.rb:
```diff
@@ -9,12 +9,37 @@ def initialize(method, path, controller, action)
+      end
+
+      def to_path_pattern(path)
+        Regexp.new("^#{path.gsub(/:([\w_]+)/, '[0-9]+')}$")
```
Почему здесь заменяем только на паттерн с цифрами? Роут-параметр может придти любым и вообще, это строка:

```
# Конфиг
/tests/:id/perform/:action

# Запрос
/tests/101/perform/start
```

psylone replied:
Стоит также помнить что регулярные выражения требуют вычислительных ресурсов.

rusdec replied:
я подумал, что это будет более элегантный способ чем тот, что я применил вначале, а именно:
```
params = {}
request_path = request_path.split('/')
path.split('/').each_with_index { |key, index| params[key] = request_path[index] if key.match(/^:/) }
```

Commit:
Исправление замечаний:
- Можно полностью сделать логику парсинга параметров в объекте роута. https://github.com/rusdec/simpler/pull/1#discussion_r176834469
- Это могут быть ещё и параметры POST-запроса. Будет лучше мерджить @request.params с роут-параметрами внутри метода params. https://github.com/rusdec/simpler/pull/1#discussion_r176834991

Commit:
Исправление замечаний:
- Вообще, часто используют _ если переменная не используется. В Go и Elixir особенно. Можно так оставить, но я бы как-то иначе назвал один из параметров. https://github.com/rusdec/simpler/pull/1#discussion_r176835451

Commit:
Исправление замечаний:
- Вот этого метода здесь по идее быть не должно. https://github.com/rusdec/simpler/pull/1/files/3dcef2186b5614fd87c4eb41c3b51b97099c8752#diff-6b549de9ab5122545fa7083788fc2ec1
- Лучше работать через case https://github.com/rusdec/simpler/pull/1#discussion_r176836334
- Здесь нарушается интерфейс. https://github.com/rusdec/simpler/pull/1/files/3dcef2186b5614fd87c4eb41c3b51b97099c8752#diff-6b549de9ab5122545fa7083788fc2ec1
- Дублируется инфа, заголовка должно быть достаточно. https://github.com/rusdec/simpler/pull/1/files/3dcef2186b5614fd87c4eb41c3b51b97099c8752#diff-6b549de9ab5122545fa7083788fc2ec1
